programs/autocrat_v0/src/lib.rs

```diff
@@ -390,39 +390,27 @@ pub mod autocrat_v0 {
     pub fn update_dao(ctx: Context<UpdateDao>, dao_params: UpdateDaoParams) -> Result<()> {
         let dao = &mut ctx.accounts.dao;
 
-        if let Some(pass_threshold_bps) = dao_params.pass_threshold_bps {
-            dao.pass_threshold_bps = pass_threshold_bps;
-        }
-
-        if let Some(base_burn_lamports) = dao_params.base_burn_lamports {
-            dao.base_burn_lamports = base_burn_lamports;
-        }
-
-        if let Some(burn_decay_per_slot_lamports) = dao_params.burn_decay_per_slot_lamports {
-            dao.burn_decay_per_slot_lamports = burn_decay_per_slot_lamports;
-        }
-
-        if let Some(slots_per_proposal) = dao_params.slots_per_proposal {
-            dao.slots_per_proposal = slots_per_proposal;
-        }
-
-        if let Some(market_taker_fee) = dao_params.market_taker_fee {
-            dao.market_taker_fee = market_taker_fee;
-        }
-
-        if let Some(twap_expected_value) = dao_params.twap_expected_value {
-            dao.twap_expected_value = twap_expected_value;
-        }
-
-        if let Some(base_lot_size) = dao_params.base_lot_size {
-            dao.base_lot_size = base_lot_size;
+        macro_rules! update_fields_if_some {
+            ($dao:expr, $dao_params:expr, $( $field:ident ),+ ) => {
+                $(
+                    if let Some(value) = $dao_params.$field {
+                        $dao.$field = value;
+                    }
+                )+
+            }
         }
 
-        if let Some(max_observation_change_per_update_lots) =
-            dao_params.max_observation_change_per_update_lots
-        {
-            dao.max_observation_change_per_update_lots = max_observation_change_per_update_lots;
-        }
+        update_fields_if_some!(
```

Comment:
The thing I don't like about this macro is that it doesn't separate out the variables on the DAO struct that can get updated from the other vars like dao and dao_parames, it all kind of blends together.

update_field_if_some!(dao, dao_params, pass_threshold_bps);
update_field_if_some!(dao, dao_params, base_burn_lamports);
update_field_if_some!(dao, dao_params, burn_decay_per_slot_lamports);

But I don't think level of copypasting is really much of better solution either.

A high effort solution I don't recommend but still mention anyway would be a proc macro that loops over all the var names in UpdateDaoParams and automatically unwraps and tries to assign them to matching var names on the Dao struct. That way you'd never have to update the testing code, it would automatically check all of the params you update. Too high effort though and needs a separate crate for the proc macro.

tl;dr LGTM!

Reply:
yeah, agree to both of above. I wish we could do a proc macro, it would be so much cleaner, but probably not worth the marginal complexity.

ended up doing something like you proposed, but without the `dao, dao_params` part: https://github.com/metaDAOproject/futarchy/pull/107/commits/b30aa39d295e351d3a9a5551d1711c1e97cde1e5